src/main/java/com/demo/classroom/Security/Filter/JwtAuthenticationFilter.java

```diff
@@ -0,0 +1,97 @@
+package com.demo.classroom.Security.Filter;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.stream.Collectors;
+
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import com.demo.classroom.Security.Service.JwtService;
+
+import jakarta.servlet.FilterChain;
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import jakarta.validation.constraints.NotNull;
+import lombok.RequiredArgsConstructor;
+
+@Component
+@RequiredArgsConstructor
+public class JwtAuthenticationFilter extends OncePerRequestFilter{
+
+    private final JwtService jwtService;
+    private final UserDetailsService userDetailsService;
+    
+
+    @Override
+    protected void doFilterInternal(
+            @NotNull HttpServletRequest request,
+            @NotNull HttpServletResponse response,
+            @NotNull FilterChain filterChain
+    ) throws ServletException, IOException {
+
+        final String authHeader = request.getHeader("Authorization");
+        final String refreshToken = request.getHeader("Refresh-Token");
+
+        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
+            filterChain.doFilter(request, response);
+            return;
+        }
+
+        try {
+            final String jwt = authHeader.substring(7); //"Bearer "=7
+            final String userName = jwtService.extractUsername(jwt);
+
+            var roles = jwtService.extractRoles(jwt);
```

Comment:
We are 100% that we are no gonna reassign `useName` variable that's why it is declared as final.

we use var on roles and userId. It doesn't matter what we use. `var` keyword is not data type specific. It will count it as a datatype what my function returns. Its an keyword from kotlin. We can change it as per requirement. No problem.

Reply:
Done